### ISSUE TITLE
Allow non-Tensor subclass in `torch.Tensor._make_wrapper_subclass`

### ISSUE THREAD

**Issue by ZhiyuanChen** [pytorch/pytorch#150861]
### 🚀 The feature, motivation and pitch

I have a NestedTensor class like this:

```python
from collections.abc import Mapping, Sequence
from typing import Any, Iterable, SupportsFloat, Tuple

import torch
from torch import Tensor

from ..utils import method_cache
from .functions import NestedTensorFuncRegistry, NestedTensorFuncWrapper
from .utils import mask_tensor, pad_tensor, tensor_mask

try:
    from typing import Self  # type: ignore[attr-defined]
except ImportError:
    from typing_extensions import Self

try:
    from torch import nested
except ImportError:
    nested = None


class NestedTensor:
    __storage: Tuple[Tensor, ...]

    dtype: torch.dtype | None = None
    device: torch.device | None = None
    requires_grad: bool | None = None
    _pin_memory: bool = False

    batch_first: bool = True
    padding_value: SupportsFloat = 0.0
    mask_value: bool = False

    def __init__(
        self,
        *tensors: Iterable[Tensor],
        dtype: torch.dtype | None = None,
        device: torch.device | None = None,
        requires_grad: bool | None = None,
        pin_memory: bool = False,
        batch_first: bool = True,
        padding_value: SupportsFloat = 0.0,
        mask_value: bool = False,
    ) -> None:
        self.dtype = dtype
        self.device = device
        self.requires_grad = requires_grad
        self._pin_memory = pin_memory
        if len(tensors) == 1 and isinstance(tensors, Sequence):
            tensors = tensors[0]  # type: ignore
        self._storage = tensors
        self.batch_first = batch_first
        self.padding_value = padding_value
        self.mask_value = mask_value

    @classmethod
    def __torch_function__(cls, func, types, args=(), kwargs=None):
        if kwargs is None:
            kwargs = {}
        if func not in NestedTensorFuncRegistry or not all(issubclass(t, (torch.Tensor, NestedTensor)) for t in types):
            args = [a.tensor if hasattr(a, "tensor") else a for a in args]
            for k, v in kwargs.items():
                if hasattr(v, "tensor"):
                    kwargs[k] = v.tensor
            return func(*args, **kwargs)
        return NestedTensorFuncRegistry[func](*args, **kwargs)

    @classmethod
    def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
        args = [a.tensor if hasattr(a, "tensor") else a for a in args]
        for k, v in kwargs.items():
            if hasattr(v, "tensor"):
                kwargs[k] = v.tensor
        return func(*args, **kwargs)

    def __getattr__(self, name: str) -> Any:
        if not self._storage:
            raise ValueError(f"Unable to get {name} from an empty {self.__class__.__name__}")
        ret = [getattr(i, name) for i in self._storage]
        elem = ret[0]
        if isinstance(elem, Tensor):
            return NestedTensor(ret, **self._state)
        if callable(elem):
            return NestedTensorFuncWrapper(ret, state=self._state)
        if elem.__hash__ is not None and len(set(ret)) == 1:
            return elem
        return ret


class NestedTensorFuncWrapper:  # pylint: disable=R0903

    __storage: Sequence[Callable] = []
    state: Mapping = {}

    def __init__(self, *callables: Iterable[Callable], state: Mapping | None = None) -> None:
        if len(callables) == 1 and isinstance(callables, Sequence):
            callables = callables[0]  # type: ignore
        self._storage = callables  # type: ignore
        if state is None:
            state = {}
        self.state = state

    @property
    def _storage(self):
        return self.__storage

    @_storage.setter
    def _storage(self, callables: Sequence):
        if not isinstance(callables, Sequence):
            raise ValueError(f"callables must be a Sequence, bug got {type(callables)}")
        if len(callables) == 0:
            raise ValueError("callables must be a non-empty Sequence.")
        if not callable(callables[0]):
            raise ValueError(f"callables must be a Sequence of Callable, bug got {type(callables[0])}")
        self.__storage = callables

    def __call__(self, *args, **kwargs) -> NestedTensor | Sequence[Tensor]:
        from .nested_tensor import NestedTensor
        from .tensor import PNTensor

        ret = [call(*args, **kwargs) for call in self._storage]
        elem = ret[0]
        if isinstance(elem, Tensor):
            try:
                return PNTensor(ret)
            except ValueError:
                return NestedTensor(ret, **self.state)
        if elem.__hash__ is not None and len(set(ret)) == 1:
            return elem
        return ret
```

Since torch's Tensor have hundreds of functions (last time I check, the `len(dir(torc.Tensor)) = 745`, and I do not want to implement them one by one. So I modify the `__getattr__` method and let it handles function calls. All I need to do is only override the methods that do not work as expected.  

But when I try to make the NestedTensor class subclass `torch.Tensor`, like this:

```

class NestedTensor(torch.Tensor):
    @staticmethod
    def __new__(
        cls,
        *tensors: Iterable[Tensor],
        dtype: torch.dtype | None = None,
        device: torch.device | None = None,
        batch_first: bool = True,
        padding_value: SupportsFloat = 0.0,
        mask_value: bool = False,
        **kwargs,
    ):
        ks = DispatchKeySet(DispatchKey.NestedTensor)
        ks = ks.add(DispatchKey.AutogradNestedTensor)
        return torch.Tensor._make_wrapper_subclass(cls, (), device=device, dtype=dtype, **kwargs)
```

All undefined methods will go through the MRO and uses `torch.Tensor`'s version. 

Which magically works, but it returns a padded tensor. This breaks some functions like `mean` and `sqrt`. 

Since there is a dedicated `torch.Tensor._make_subclass` for subclassing Tensor, it would be reasonable to make `torch.Tensor._make_wrapper_subclass` works for non-`Tensor` subclass. 

By the way, the second parameter of `torch.Tensor._make_wrapper_subclass` (size) is also confusing, I passed an empty tuple as of now, but it seems to work. 

### Alternatives

Introduce another API that works for non-tensor subclasses. 

### Additional context

You can find a MWE of the two versions at [tensor1](https://github.com/ZhiyuanChen/DanLing/blob/tensor/danling/tensor/nested_tensor.py) and [tensor2](https://github.com/ZhiyuanChen/DanLing/blob/tensor2/danling/tensor/nested_tensor.py).

**Comment by albanD** [comment]
> Since there is a dedicated torch.Tensor._make_subclass for subclassing Tensor, it would be reasonable to make torch.Tensor._make_wrapper_subclass works for non-Tensor subclass.

I'm afraid this is not technically possible. A wrapper subclass is a torch_dispatch class (meaning it's handled in C++). But for your object to be able to live in c++, we need it to be a real Tensor. Hence being a subclass of Tensor is a must-have to be a wrapper subclass.

> By the way, the second parameter of torch.Tensor._make_wrapper_subclass (size) is also confusing, I passed an empty tuple as of now, but it seems to work.

That is true for all Tensors, empty size tuple means it is a "scalar tensor" that contains only a single element. All factory functions support this.

**Comment by ZhiyuanChen** [comment]
> I'm afraid this is not technically possible. A wrapper subclass is a torch_dispatch class (meaning it's handled in C++). But for your object to be able to live in c++, we need it to be a real Tensor. Hence being a subclass of Tensor is a must-have to be a wrapper subclass.

Is it possible to have another base class that tells torch its a torch_dispatch class, but does not have all the Tensor's methods? It would literally take ages to override every methods. It also makes it really hard to maintain. 

Currently, 

> That is true for all Tensors, empty size tuple means it is a "scalar tensor" that contains only a single element. All factory functions support this.

Ahh right! 
Can I ask what happens if I just pass an empty tuple? The size is computed internally in the NestedTensor class (so after `__init__` call, but I guess it won't be difficult to move it here. 

By the way, I noticed that the torch's built-in NestedTensor have significant improvements in the past versions. But I think it has too much restrictions. For example, it only supports one variable dimension (I'm not 100% sure, but when I pass a nested image tensor to F.linear, it complains). 
Many other applications (like structure prediction for biomolecules) require more than 1 dimension.

**Comment by function2-llx** [comment]
Not sure if this would be helpful. It seems that literally hundreds of Tensor methods has been implemented for `MaskedTensor`. Following a similar design pattern may significantly reduce the maintenance efforts.
https://github.com/pytorch/pytorch/blob/68034198e5c25f10005d13c6a68cb7a4e3a76f1e/torch/masked/maskedtensor/_ops_refs.py
